package/src/nvidiamon.cpp

```diff
@@ -0,0 +1,207 @@
+// Copyright (C) CERN, 2020
+
+#include "nvidiamon.h"
+
+#include <stdlib.h>
+#include <unistd.h>
+
+#include <iostream>
+#include <sstream>
+#include <string>
+#include <unordered_map>
+
+#include "utils.h"
+
+// This is not very nice, but we don't want to add more special runtime
+// flags, pending a proper decision on logging (#106)
+#define NVIDIA_DEBUG 0
+
+// Constructor; uses RAII pattern to be valid after construction
+nvidiamon::nvidiamon()
+    : nvidia_params{prmon::default_nvidia_params},
+      nvidia_stats{},
+      nvidia_average_stats{},
+      nvidia_total_stats{},
+      iterations{0L} {
+  for (const auto& nvidia_param : nvidia_params) {
+    nvidia_stats[nvidia_param] = 0;
+    nvidia_peak_stats[nvidia_param] = 0;
+    nvidia_average_stats[nvidia_param] = 0;
+    nvidia_total_stats[nvidia_param] = 0;
+  }
+
+  // Attempt to execute nvidia-smi
+  // If this works we are valid, but if not then we can't get data
+  valid = test_nvidia_smi();
+}
+
+void nvidiamon::update_stats(const std::vector<pid_t>& pids) {
+  const std::vector<std::string> cmd = {"nvidia-smi", "pmon", "-s",
+                                        "um",         "-c",   "1"};
+
+  for (auto& stat : nvidia_stats) stat.second = 0;
+
+  auto cmd_result = prmon::cmd_pipe_output(cmd);
+  if (cmd_result.first) {
+    // Failed
+    return;
+  }
+
+  if (NVIDIA_DEBUG) {
+    std::cout << "nvidiamon::update_stats got the following output ("
+              << cmd_result.second.size() << "): " << std::endl;
+    int i = 0;
+    for (const auto& s : cmd_result.second) {
+      std::cout << i << " -> " << s << std::endl;
+      ++i;
+    }
+  }
+
+  // Loop over output
+  unsigned int gpu_idx{}, sm{}, mem{}, enc{}, dec{}, fb_mem{};
+  pid_t pid{};
+  std::string cg_type{}, cmd_name{};
+  std::unordered_map<unsigned int, bool>
+      activegpus{};  // Avoid double counting active GPUs
+  for (const auto& s : cmd_result.second) {
+    if (s[0] != '#') continue;
+    std::istringstream instr(s);
+    instr >> gpu_idx >> pid >> cg_type >> sm >> mem >> enc >> dec >> fb_mem >>
+        cmd_name;
+    auto read_ok = !(instr.fail() || instr.bad());  // eof() is ok
+    if (read_ok) {
+      if (NVIDIA_DEBUG) {
+        std::cout << "Good read: " << gpu_idx << " " << pid << " " << cg_type
+                  << " " << sm << " " << mem << " " << enc << " " << dec << " "
+                  << fb_mem << " " << cmd_name << std::endl;
+      }
+
+      // Filter on PID value, so we only add stats for our processes
+      for (auto const p : pids) {
+        if (p == pid) {
+          nvidia_stats["gpusmpct"] += sm;
+          nvidia_stats["gpumempct"] += mem;
+          nvidia_stats["gpufbmem"] += fb_mem;
```

Comment:
It seems as if we're still accumulating memory in MB, we probably need:

```suggestion
          nvidia_stats["gpufbmem"] += fb_mem * 1024;
```

to convert this into KB to be consistent.

Reply:
Yes, this is correct

Reply:
In fact, better to setup a constant that describes what we are doing than use a magic number